File: nemo_retriever/src/nemo_retriever/service/policy.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES.
# All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Server-side trust boundary for per-request pipeline overrides.

The :class:`~nemo_retriever.service.models.pipeline_spec.PipelineSpec`
shipped by a client cannot be applied blindly: it would let a tenant
redirect NIM traffic, exfiltrate to arbitrary HTTP endpoints, or write
files to attacker-chosen paths. This module centralises that
trust-boundary logic so the router, worker, and ``pipeline-config``
introspection endpoint all see the same picture of what is allowed.

Two layers:

1. **Denylist of endpoint / api_key fields**. The server *always* owns
   these — the only way to influence them is to update
   ``ServiceConfig.nim_endpoints``.
2. **Per-stage allowlist of param keys**. Defaults to a conservative set
   of "shape" fields (chunk sizes, batch sizes, output flags). Operators
   can widen the allowlist via the new ``pipeline_overrides`` section of
   ``retriever-service.yaml``.

The module is pure-Python: it takes plain dicts (the wire format) and
returns plain dicts (the validated, server-merged form). No FastAPI
imports, so unit tests can exercise it directly.
"""

from __future__ import annotations

from typing import Any

from nemo_retriever.service.models.pipeline_spec import PipelineSpec


# ----------------------------------------------------------------------
# Trust-sensitive field names — never accept these from a client
# ----------------------------------------------------------------------

_DENYLIST_KEY_SUBSTRINGS: tuple[str, ...] = (
    "invoke_url",
    "endpoint",
    "endpoint_url",
    "api_key",
    "api_base",
    "auth_token",
    "function_id",
    "audio_endpoints",
    "remote_invoke",
    "callback_url",
    "lancedb_uri",
    "storage_uri",
    "vectordb_url",
    "embed_invoke_url",
    "embedding_endpoint",
    "page_elements_invoke_url",
    "ocr_invoke_url",
    "table_structure_invoke_url",
    "graphic_elements_invoke_url",
    "nemotron_parse_invoke_url",
)


def _is_trust_sensitive(key: str) -> bool:
    """Return ``True`` when ``key`` matches a denied substring."""
    lower = key.lower()
    return any(needle in lower for needle in _DENYLIST_KEY_SUBSTRINGS)


# ----------------------------------------------------------------------
# Default per-stage allowlist — overridable from YAML
# ----------------------------------------------------------------------

_DEFAULT_ALLOWED_EXTRACT_KEYS: frozenset[str] = frozenset(
    {
        "extract_text",
        "extract_images",
        "extract_tables",
        "extract_charts",
        "extract_infographics",
        "extract_page_as_image",
        "method",
        "use_table_structure",
        "table_output_format",
        "use_graphic_elements",
        "dpi",
        "image_format",
        "jpeg_quality",
        "render_mode",
        "inference_batch_size",
        "ocr_version",
        "output_column",
        "num_detections_column",
        "counts_by_label_column",
    }
)

_DEFAULT_ALLOWED_EMBED_KEYS: frozenset[str] = frozenset(
    {
        "input_type",
        "embed_modality",
        "embed_granularity",
        "text_elements_modality",
        "structured_elements_modality",
        "text_column",
        "inference_batch_size",
        "output_column",
        "embedding_dim_column",
        "has_embedding_column",
        "embed_output_column",
        "embed_inference_batch_size",
        "query_max_length",
        "dimensions",
    }
)

_DEFAULT_ALLOWED_DEDUP_KEYS: frozenset[str] = frozenset({"content_hash", "bbox_iou", "iou_threshold"})

_DEFAULT_ALLOWED_SPLIT_KEYS: frozenset[str] = frozenset(
    {"max_tokens", "overlap_tokens", "tokenizer_model_id", "encoding"}
)

# Stages we explicitly disallow in Phase 1. Each is wired up in a later
# phase under its own configuration.
_PHASE_1_STAGES: frozenset[str] = frozenset({"extract", "dedup", "embed", "filter"})


class PolicyError(ValueError):
    """Raised by :func:`validate_pipeline_spec` when a client spec is rejected.

    Carries an HTTP-friendly *status_code* (400 for malformed input, 403
    for a policy denial) and a human-readable *detail* string suitable
    for the FastAPI exception handler.
    """

    def __init__(self, detail: str, *, status_code: int = 400) -> None:
        super().__init__(detail)
        self.status_code = status_code
        self.detail = detail


# ----------------------------------------------------------------------
# Public configuration object — mirrors ServiceConfig.pipeline_overrides
# ----------------------------------------------------------------------


class PipelineOverridesPolicy:
    """Operator-tunable view of what client overrides are admissible.

    Operators set this via ``pipeline_overrides`` in ``retriever-service.yaml``;
    callers in the router/worker get a single :class:`PipelineOverridesPolicy`
    instance from ``ServiceConfig`` and pass it to :func:`validate_pipeline_spec`.
    """

    def __init__(
        self,
        *,
        mode: str = "allow_list",
        allowed_stages: frozenset[str] | None = None,
        extra_extract_keys: frozenset[str] = frozenset(),
        extra_embed_keys: frozenset[str] = frozenset(),
        extra_dedup_keys: frozenset[str] = frozenset(),
        extra_split_keys: frozenset[str] = frozenset(),
    ) -> None:
        if mode not in {"reject", "allow_list", "allow_all"}:
            raise ValueError(
                f"pipeline_overrides.mode must be one of 'reject', 'allow_list', 'allow_all'; got {mode!r}"
            )
        self.mode = mode
        self.allowed_stages = allowed_stages if allowed_stages is not None else _PHASE_1_STAGES
        self.allowed_extract_keys = _DEFAULT_ALLOWED_EXTRACT_KEYS | extra_extract_keys
        self.allowed_embed_keys = _DEFAULT_ALLOWED_EMBED_KEYS | extra_embed_keys
        self.allowed_dedup_keys = _DEFAULT_ALLOWED_DEDUP_KEYS | extra_dedup_keys
        self.allowed_split_keys = _DEFAULT_ALLOWED_SPLIT_KEYS | extra_split_keys

    def describe(self) -> dict[str, Any]:
        """Render the policy as a JSON-safe dict for the introspection endpoint."""
        return {
            "mode": self.mode,
            "allowed_stages": sorted(self.allowed_stages),
            "allowed_extract_keys": sorted(self.allowed_extract_keys),
            "allowed_embed_keys": sorted(self.allowed_embed_keys),
            "allowed_dedup_keys": sorted(self.allowed_dedup_keys),
            "allowed_split_keys": sorted(self.allowed_split_keys),
            "denied_key_substrings": sorted(_DENYLIST_KEY_SUBSTRINGS),
        }


# ----------------------------------------------------------------------
# Validation entry point
# ----------------------------------------------------------------------


def _scrub_trust_sensitive(params: dict[str, Any] | None, stage: str) -> dict[str, Any] | None:
    """Strip trust-sensitive keys from a params dict, raising on hit.

    We *reject* (rather than silently drop) so the client gets a clear
    error and knows their request will not behave as written.
    """
    if params is None:
        return None
    bad = [k for k in params if _is_trust_sensitive(k)]
    if bad:
        raise PolicyError(
            f"{stage}_params: rejected trust-sensitive keys {bad!r}. "
            "Endpoint URLs and API keys are configured via the server's "
            "nim_endpoints section and cannot be overridden per-request.",
            status_code=403,
        )
    return params


def _enforce_allowlist(
    params: dict[str, Any] | None,
    allowed: frozenset[str],
    stage: str,
    *,
    mode: str,
) -> dict[str, Any] | None:
    """Apply per-stage allowlist enforcement based on the policy mode."""
    if params is None or mode == "allow_all":
        return params
    extras = [k for k in params if k not in allowed]
    if not extras:
        return params
    if mode == "reject":
        raise PolicyError(
            f"{stage}_params: client overrides are disabled (pipeline_overrides.mode='reject'). "
            f"Offending keys: {extras!r}.",
            status_code=403,
        )
    raise PolicyError(
        f"{stage}_params: keys {extras!r} are not in the allow_list. "
        "Ask the service operator to widen pipeline_overrides.extra_*_keys "
        "or remove them from your request.",
        status_code=403,
    )


def validate_pipeline_spec(
    spec: PipelineSpec | None,
    policy: PipelineOverridesPolicy,
) -> PipelineSpec | None:
    """Return a sanitized copy of *spec* or raise :class:`PolicyError`.

    Returns ``None`` when the spec is missing or empty (so the worker can
    short-circuit to the legacy startup-baked pipeline).
    """
    if spec is None or spec.is_empty():
        return None

    if policy.mode == "reject":
        raise PolicyError(
            "Per-request pipeline overrides are disabled on this service "
            "(pipeline_overrides.mode='reject'). Update retriever-service.yaml "
            "to enable them.",
            status_code=403,
        )

    for stage_name in spec.stage_order:
        if stage_name not in policy.allowed_stages:
            raise PolicyError(
                f"stage {stage_name!r} is not in pipeline_overrides.allowed_stages. "
                f"Allowed in this phase: {sorted(policy.allowed_stages)}.",
                status_code=403,
            )

    # Reject groups B/C/D stages outright until their phase lands.
    for forbidden in ("caption_params", "store_params", "vdb_upload_params", "webhook_params"):
        if getattr(spec, forbidden) is not None:
            raise PolicyError(
                f"{forbidden} overrides are not yet supported in service run_mode "
                "(scheduled for a future phase). Configure the corresponding "
                "stage via retriever-service.yaml instead.",
                status_code=501,
            )

    # Endpoint/API-key denylist applies to every params block, even ones
    # we currently accept — defense in depth in case a new field name
    # slips into the allowlist without being audited.
    _scrub_trust_sensitive(spec.extract_params, "extract")
    _scrub_trust_sensitive(spec.embed_params, "embed")
    _scrub_trust_sensitive(spec.dedup_params, "dedup")
    _scrub_trust_sensitive(spec.split_config, "split")

    _enforce_allowlist(spec.extract_params, policy.allowed_extract_keys, "extract", mode=policy.mode)
    _enforce_allowlist(spec.embed_params, policy.allowed_embed_keys, "embed", mode=policy.mode)
    _enforce_allowlist(spec.dedup_params, policy.allowed_dedup_keys, "dedup", mode=policy.mode)
    if spec.split_config is not None:
        for source_type, cfg in spec.split_config.items():
            if not isinstance(cfg, dict):
                continue
            _enforce_allowlist(cfg, policy.allowed_split_keys, f"split[{source_type}]", mode=policy.mode)

    return spec

File: nemo_retriever_rust/crates/nrr-core/src/policy.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES.
// All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Pipeline-overrides policy.
//!
//! Mirrors `nemo_retriever.service.policy`:
//! * `reject` — every override is denied.
//! * `allow_list` — only the built-in audited keys plus operator-supplied
//!   `extra_*_keys` are accepted. Endpoint URLs and API keys are ALWAYS denied.
//! * `allow_all` — every key passes except the endpoint/api_key denylist.
//!
//! Sink endpoints (image store / webhook / vdb_upload) are gated by URL/scheme
//! allow-lists in [`SinkUrlAllowlist`].

use std::collections::HashSet;

use serde_json::Value as JsonValue;
use thiserror::Error;

use crate::config::OverridesMode;
use crate::models::PipelineSpec;

/// Keys that NEVER ride a per-request override regardless of policy mode.
/// They are entirely server-owned (NIM endpoint URLs, API keys, model IDs).
const TRUST_OWNED_EXTRACT_KEYS: &[&str] = &[
    "invoke_url",
    "api_key",
    "page_elements_invoke_url",
    "page_elements_api_key",
    "ocr_invoke_url",
    "ocr_api_key",
    "graphic_elements_invoke_url",
    "table_structure_invoke_url",
    "nemotron_parse_invoke_url",
];
const TRUST_OWNED_EMBED_KEYS: &[&str] =
    &["embed_invoke_url", "embedding_endpoint", "api_key"];
const TRUST_OWNED_CAPTION_KEYS: &[&str] = &["endpoint_url", "api_key", "model_name"];

/// Default audited "shape" keys per stage. These are the knobs operators
/// have already approved for tenant override (chunk sizes, batch sizes,
/// output column names …). Anything else is denied unless `allow_all`
/// is in effect.
const DEFAULT_EXTRACT_KEYS: &[&str] = &[
    "extract_text",
    "extract_charts",
    "extract_tables",
    "extract_infographics",
    "text_depth",
    "chunk_size",
    "chunk_overlap",
];
const DEFAULT_EMBED_KEYS: &[&str] = &[
    "embed_modality",
    "inference_batch_size",
    "input_type",
    "nim_http_max_concurrent",
    "text_column",
    "output_column",
];
const DEFAULT_DEDUP_KEYS: &[&str] = &["enabled", "method", "threshold"];
const DEFAULT_SPLIT_KEYS: &[&str] = &["pages_per_chunk", "max_pages"];
const DEFAULT_STORE_KEYS: &[&str] = &["uri", "compression"];
const DEFAULT_WEBHOOK_KEYS: &[&str] = &["url", "headers", "method"];
const DEFAULT_VDB_UPLOAD_KEYS: &[&str] = &[
    "lancedb_uri",
    "table_name",
    "overwrite",
    "vector_dim",
    "meta_dataframe_id",
];
const DEFAULT_CAPTION_KEYS: &[&str] = &[
    "prompt",
    "system_prompt",
    "batch_size",
    "max_tokens",
    "temperature",
];

#[derive(Debug, Clone)]
pub struct SinkUrlAllowlist {
    pub storage_uri_schemes: Vec<String>,
    pub webhook_url_prefixes: Vec<String>,
    pub vdb_uri_schemes: Vec<String>,
}

impl SinkUrlAllowlist {
    fn check_uri_scheme(uri: &str, schemes: &[String]) -> bool {
        if schemes.iter().any(|s| s == "*") {
            return true;
        }
        schemes.iter().any(|s| uri.starts_with(s.as_str()))
    }

    pub fn allow_storage(&self, uri: &str) -> bool {
        Self::check_uri_scheme(uri, &self.storage_uri_schemes)
    }
    pub fn allow_webhook(&self, url: &str) -> bool {
        if self.webhook_url_prefixes.iter().any(|p| p == "*") {
            return true;
        }
        self.webhook_url_prefixes
            .iter()
            .any(|p| url.starts_with(p.as_str()))
    }
    pub fn allow_vdb_upload(&self, uri: &str) -> bool {
        Self::check_uri_scheme(uri, &self.vdb_uri_schemes)
    }
}

#[derive(Debug, Clone)]
pub struct PipelineOverridesPolicy {
    pub mode: OverridesMode,
    pub extra_extract_keys: HashSet<String>,
    pub extra_embed_keys: HashSet<String>,
    pub extra_dedup_keys: HashSet<String>,
    pub extra_split_keys: HashSet<String>,
    pub extra_store_keys: HashSet<String>,
    pub extra_webhook_keys: HashSet<String>,
    pub extra_vdb_upload_keys: HashSet<String>,
    pub extra_caption_keys: HashSet<String>,
    pub sinks: SinkUrlAllowlist,
    pub caption_enabled: bool,
}

impl PipelineOverridesPolicy {
    pub fn describe(&self) -> serde_json::Value {
        serde_json::json!({
            "mode": match self.mode {
                OverridesMode::Reject => "reject",
                OverridesMode::AllowList => "allow_list",
                OverridesMode::AllowAll => "allow_all",
            },
            "extract_keys": merged_keys(DEFAULT_EXTRACT_KEYS, &self.extra_extract_keys),
            "embed_keys": merged_keys(DEFAULT_EMBED_KEYS, &self.extra_embed_keys),
            "dedup_keys": merged_keys(DEFAULT_DEDUP_KEYS, &self.extra_dedup_keys),
            "split_keys": merged_keys(DEFAULT_SPLIT_KEYS, &self.extra_split_keys),
            "store_keys": merged_keys(DEFAULT_STORE_KEYS, &self.extra_store_keys),
            "webhook_keys": merged_keys(DEFAULT_WEBHOOK_KEYS, &self.extra_webhook_keys),
            "vdb_upload_keys": merged_keys(DEFAULT_VDB_UPLOAD_KEYS, &self.extra_vdb_upload_keys),
            "caption_keys": merged_keys(DEFAULT_CAPTION_KEYS, &self.extra_caption_keys),
            "caption_enabled": self.caption_enabled,
            "sinks": {
                "storage_uri_schemes": &self.sinks.storage_uri_schemes,
                "webhook_url_prefixes": &self.sinks.webhook_url_prefixes,
                "vdb_uri_schemes": &self.sinks.vdb_uri_schemes,
            }
        })
    }
}

fn merged_keys(defaults: &[&str], extras: &HashSet<String>) -> Vec<String> {
    let mut out: Vec<String> = defaults.iter().map(|s| (*s).to_string()).collect();
    out.extend(extras.iter().cloned());
    out.sort();
    out.dedup();
    out
}

#[derive(Debug, Error)]
pub enum PolicyError {
    #[error("forbidden override key: {key}")]
    ForbiddenKey { key: String, status_code: u16 },
    #[error("override mode is 'reject'; client overrides are not allowed")]
    Rejected { status_code: u16 },
    #[error("forbidden sink uri/url: {uri}")]
    ForbiddenSink { uri: String, status_code: u16 },
    #[error("caption stage requested but caption_invoke_url is not configured")]
    CaptionDisabled { status_code: u16 },
    #[error("malformed pipeline spec: {0}")]
    Malformed(String),
}

impl PolicyError {
    pub fn status_code(&self) -> u16 {
        match self {
            PolicyError::ForbiddenKey { status_code, .. } => *status_code,
            PolicyError::Rejected { status_code } => *status_code,
            PolicyError::ForbiddenSink { status_code, .. } => *status_code,
            PolicyError::CaptionDisabled { status_code } => *status_code,
            PolicyError::Malformed(_) => 400,
        }
    }
}

/// Validate a client-supplied [`PipelineSpec`] against the policy.
///
/// Returns the spec unchanged on success. The server merges its own
/// trust-owned defaults (URLs / API keys) on top of the returned spec at
/// pipeline build time, so the validated spec only carries "shape" knobs.
pub fn validate_pipeline_spec(
    spec: &PipelineSpec,
    policy: &PipelineOverridesPolicy,
) -> Result<PipelineSpec, PolicyError> {
    if spec.is_empty() {
        return Ok(spec.clone());
    }

    if matches!(policy.mode, OverridesMode::Reject) {
        return Err(PolicyError::Rejected { status_code: 403 });
    }

    let mode = policy.mode;

    check_params_block(
        spec.extract_params.as_ref(),
        DEFAULT_EXTRACT_KEYS,
        &policy.extra_extract_keys,
        TRUST_OWNED_EXTRACT_KEYS,
        mode,
    )?;
    check_params_block(
        spec.embed_params.as_ref(),
        DEFAULT_EMBED_KEYS,
        &policy.extra_embed_keys,
        TRUST_OWNED_EMBED_KEYS,
        mode,
    )?;
    check_params_block(
        spec.dedup_params.as_ref(),
        DEFAULT_DEDUP_KEYS,
        &policy.extra_dedup_keys,
        &[],
        mode,
    )?;
    check_params_block(
        spec.split_config.as_ref(),
        DEFAULT_SPLIT_KEYS,
        &policy.extra_split_keys,
        &[],
        mode,
    )?;
    check_params_block(
        spec.vdb_upload_params.as_ref(),
        DEFAULT_VDB_UPLOAD_KEYS,
        &policy.extra_vdb_upload_keys,
        &[],
        mode,
    )?;

    if let Some(store) = &spec.store_params {
        check_params_block(
            Some(store),
            DEFAULT_STORE_KEYS,
            &policy.extra_store_keys,
            &[],
            mode,
        )?;
        if let Some(uri) = store.get("uri").and_then(|v| v.as_str()) {
            if !policy.sinks.allow_storage(uri) {
                return Err(PolicyError::ForbiddenSink {
                    uri: uri.into(),
                    status_code: 403,
                });
            }
        }
    }

    if let Some(webhook) = &spec.webhook_params {
        check_params_block(
            Some(webhook),
            DEFAULT_WEBHOOK_KEYS,
            &policy.extra_webhook_keys,
            &[],
            mode,
        )?;
        if let Some(url) = webhook.get("url").and_then(|v| v.as_str()) {
            if !policy.sinks.allow_webhook(url) {
                return Err(PolicyError::ForbiddenSink {
                    uri: url.into(),
                    status_code: 403,
                });
            }
        }
    }

    if let Some(vdb) = &spec.vdb_upload_params {
        if let Some(uri) = vdb.get("lancedb_uri").and_then(|v| v.as_str()) {
            if !policy.sinks.allow_vdb_upload(uri) {
                return Err(PolicyError::ForbiddenSink {
                    uri: uri.into(),
                    status_code: 403,
                });
            }
        }
    }

    if spec.caption_params.is_some() {
        if !policy.caption_enabled {
            return Err(PolicyError::CaptionDisabled { status_code: 403 });
        }
        check_params_block(
            spec.caption_params.as_ref(),
            DEFAULT_CAPTION_KEYS,
            &policy.extra_caption_keys,
            TRUST_OWNED_CAPTION_KEYS,
            mode,
        )?;
    }

    Ok(spec.clone())
}

fn check_params_block(
    params: Option<&JsonValue>,
    defaults: &[&str],
    extras: &HashSet<String>,
    denied: &[&str],
    mode: OverridesMode,
) -> Result<(), PolicyError> {
    let Some(JsonValue::Object(map)) = params else {
        return Ok(());
    };
    for (k, _) in map.iter() {
        if denied.iter().any(|d| *d == k.as_str()) {
            return Err(PolicyError::ForbiddenKey {
                key: k.clone(),
                status_code: 403,
            });
        }
        if matches!(mode, OverridesMode::AllowAll) {
            continue;
        }
        let allowed = defaults.iter().any(|d| *d == k.as_str()) || extras.contains(k);
        if !allowed {
            return Err(PolicyError::ForbiddenKey {
                key: k.clone(),
                status_code: 403,
            });
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn policy(mode: OverridesMode) -> PipelineOverridesPolicy {
        PipelineOverridesPolicy {
            mode,
            extra_extract_keys: HashSet::new(),
            extra_embed_keys: HashSet::new(),
            extra_dedup_keys: HashSet::new(),
            extra_split_keys: HashSet::new(),
            extra_store_keys: HashSet::new(),
            extra_webhook_keys: HashSet::new(),
            extra_vdb_upload_keys: HashSet::new(),
            extra_caption_keys: HashSet::new(),
            sinks: SinkUrlAllowlist {
                storage_uri_schemes: vec![],
                webhook_url_prefixes: vec![],
                vdb_uri_schemes: vec![],
            },
            caption_enabled: false,
        }
    }

    #[test]
    fn reject_mode_blocks_overrides() {
        let p = policy(OverridesMode::Reject);
        let mut spec = PipelineSpec::default();
        spec.embed_params = Some(json!({ "inference_batch_size": 32 }));
        let err = validate_pipeline_spec(&spec, &p).unwrap_err();
        assert_eq!(err.status_code(), 403);
    }

    #[test]
    fn allow_list_blocks_endpoint_keys() {
        let p = policy(OverridesMode::AllowList);
        let mut spec = PipelineSpec::default();
        spec.embed_params = Some(json!({ "embed_invoke_url": "http://attacker" }));
        let err = validate_pipeline_spec(&spec, &p).unwrap_err();
        assert_eq!(err.status_code(), 403);
    }

    #[test]
    fn allow_list_accepts_default_keys() {
        let p = policy(OverridesMode::AllowList);
        let mut spec = PipelineSpec::default();
        spec.embed_params = Some(json!({ "inference_batch_size": 64 }));
        validate_pipeline_spec(&spec, &p).unwrap();
    }

    #[test]
    fn allow_all_still_blocks_endpoint_keys() {
        let p = policy(OverridesMode::AllowAll);
        let mut spec = PipelineSpec::default();
        spec.embed_params = Some(json!({ "embed_invoke_url": "http://attacker" }));
        let err = validate_pipeline_spec(&spec, &p).unwrap_err();
        assert_eq!(err.status_code(), 403);
    }
}
